File: yurt-convention/src/main/java/com/github/liulus/yurt/convention/util/Pages.java
package com.github.liulus.yurt.convention.util;

import com.github.liulus.yurt.convention.data.DefaultPage;
import com.github.liulus.yurt.convention.data.Page;
import com.github.liulus.yurt.convention.data.PageList;
import com.github.liulus.yurt.convention.data.Pageable;
import org.jetbrains.annotations.NotNull;

import java.util.List;

/**
 * @author liulu
 * @version V1.0
 * @since 2020/11/9
 */
public abstract class Pages {

    @NotNull
    public static final Page<?> EMPTY = new DefaultPage<>();

    @NotNull
    public static <T> Page<T> page(@NotNull List<T> results) {
        if (results instanceof PageList) {
            PageList<T> ts = (PageList<T>) results;
            return page(ts.getPageNum(), ts.getPageSize(), ts, ts.getTotalRecords());
        }


        return new DefaultPage<>(results);
    }

    @NotNull
    public static <T> Page<T> page(Pageable pageable, List<T> results, long totalRecords) {
        return new DefaultPage<>(pageable.getPageNum(), pageable.getPageSize(), results, totalRecords);
    }

    @NotNull
    public static <T> Page<T> page(int pageNum, int pageSize, List<T> results, long totalRecords) {
        return new DefaultPage<>(pageNum, pageSize, results, totalRecords);
    }


}

File: yurt-convention-api/src/main/kotlin/com/github/liulus/yurt/convention/page/Pages.kt
package com.github.liulus.yurt.convention.page

/**
 *
 * @author liulu
 * @version V1.0
 * @since 2020/10/29
 */
object Pages {

    @JvmField
    val EMPTY: Page<Any> = DefaultPage()

    @JvmStatic
    fun <T> page(pageable: Pageable, results: List<T>, totalRecords: Int): Page<T> {
        return DefaultPage(pageable, results, totalRecords)
    }

    @JvmStatic
    fun <T, S> page(pageable: Pageable, results: List<T>, totalCount: Int, transformer: (T) -> S): Page<S>? {
        return page(pageable, results.asSequence().map(transformer).toList(), totalCount)
    }

    @JvmStatic
    fun <T> page(pageNum: Int, pageSize: Int, results: List<T>, totalRecords: Int): Page<T> {
        return DefaultPage(pageNum, pageSize, results, totalRecords)
    }

}